src/lsp/cobol_indent/indenter.ml

```diff
@@ -62,22 +72,37 @@ let indenter ~source_format (str:string) (rdl:indent_record list) range =
   String.concat "\n" strl
 
 (*indent a range of file, with the default indent_config*)
-let indent_range' ~dialect ~source_format ~range ~filename ~contents =
+let indent_range ~dialect ~source_format ~range ~filename ~contents =
+  let src_format =
+    (* Note: this value doesn't actually matter, it will be overriden
+       immediately by [fold_source_lines] calling [on_initial_source_format]
+       below. *)
+    match source_format with
+    | Cobol_config.Auto -> Cobol_preproc.Src_format.from_config SFFixed
+    | SF source_format -> Cobol_preproc.Src_format.from_config source_format
+  in
   let state =
     Cobol_preproc.fold_source_lines ~dialect ~source_format
+      ~on_initial_source_format:(fun src_format st -> { st with src_format })
+      ~on_compiler_directive:(fun _ { payload = cd;  _} st ->
+        match cd with
+        | CDirSource { payload = src_format; _ } -> { st with src_format }
+        | _ -> st
+      )
       ~skip_compiler_directives_text:true
       ~f:(fun _lnum line acc -> Indent_check.check_indentation line acc)
       (String { filename; contents })
-      { scope = BEGIN; context  = []; acc = []; range }
+      { src_format; scope = BEGIN; context  = []; acc = []; range }
   in
   (* NB: note here we ignore diagnostics *)
   let ind_recds = state.result.acc in
-  indenter ~source_format contents ind_recds state.result.range
+  indenter
+    ~source_format:state.result.src_format contents ind_recds state.result.range
```

Comment:
Note, here it is the source format at the end of the source text that you are providing to `indenter`. Not sure this is intended (and what `indenter` uses it for).

Reply:
`indenter` uses this to know whether to indent or not :)

The indenter works in two passes — first it computes the changes using `fold_source_lines`, then it applies the changes separately. The first pass doesn't care about the format; the second pass uses the format to determine how to indent (e.g. ignore the first chars of the line).

I think the design is not ideal, because we end up with a single huge modification to the text where VSCode guidelines [state](https://code.visualstudio.com/api/language-extensions/programmatic-language-features):

> You should always return the smallest possible text edits that result in the source code being formatted. This is crucial to ensure that markers such as diagnostic results are adjusted correctly and are not lost.

It also assumes that there is a single source format per file, and I followed that assumption here -- this is why I use the source format at the end of the source text, so that the compiler directives are taken into account. If that assumption is wrong, which your reply makes me think it might be, some additional refactoring will be needed to deal with that.